File: Sources/RewordMeCore/ModelSelection.swift
import Foundation

/// Picks the default model for a provider: the least costly tier,
/// preferring stable releases, newest first.
public enum ModelSelection {
    public static func defaultModel(for kind: ProviderKind, from models: [ModelInfo]) -> ModelInfo? {
        guard !models.isEmpty else { return nil }
        let lowestTier = models.map { costTier(kind: kind, id: $0.id) }.min()!
        let candidates = models.filter { costTier(kind: kind, id: $0.id) == lowestTier }
        let stable = candidates.filter { !isPreview($0.id) }
        let pool = stable.isEmpty ? candidates : stable
        return pool.max { $0.id < $1.id }
    }

    /// Lower tier = cheaper. Unknown names land mid-tier so a cheap
    /// family keyword always wins when present.
    static func costTier(kind: ProviderKind, id: String) -> Int {
        let lower = id.lowercased()
        switch kind {
        case .anthropic:
            if lower.contains("haiku") { return 0 }
            if lower.contains("sonnet") { return 1 }
            if lower.contains("opus") { return 2 }
            if lower.contains("fable") || lower.contains("mythos") { return 3 }
            return 2
        case .openai:
            if lower.contains("nano") { return 0 }
            if lower.contains("mini") { return 1 }
            return 2
        case .gemini:
            if lower.contains("flash") && lower.contains("lite") { return 0 }
            if lower.contains("flash") { return 1 }
            if lower.contains("pro") { return 2 }
            return 3
        case .mistral:
            if lower.contains("ministral") || lower.contains("tiny") { return 0 }
            if lower.contains("nemo") || lower.contains("small") { return 1 }
            if lower.contains("medium") { return 2 }
            if lower.contains("large") { return 3 }
            return 2
        case .xai:
            if lower.contains("mini") { return 0 }
            if lower.contains("fast") { return 1 }
            return 2
        case .deepseek:
            if lower.contains("chat") { return 0 }
            return 1
        }
    }

    static func isPreview(_ id: String) -> Bool {
        let lower = id.lowercased()
        return lower.contains("preview") || lower.contains("-exp") || lower.contains("latest")
    }
}

File: kmp/core/domain/src/commonMain/kotlin/dev/mjablonski/rewordme/domain/ModelSelection.kt
package dev.mjablonski.rewordme.domain

import dev.mjablonski.rewordme.models.ModelInfo
import dev.mjablonski.rewordme.models.ProviderKind

/**
 * Picks the default model for a provider: the least costly tier,
 * preferring stable releases, newest first.
 */
object ModelSelection {
    fun defaultModel(kind: ProviderKind, models: List<ModelInfo>): ModelInfo? {
        if (models.isEmpty()) return null
        // Local models are all free; Ollama lists the most recently
        // pulled/updated model first, which is the best default.
        if (kind.access != dev.mjablonski.rewordme.models.ProviderAccess.API_KEY) return models.first()

        val lowestTier = models.minOf { costTier(kind, it.id) }
        val candidates = models.filter { costTier(kind, it.id) == lowestTier }
        val stable = candidates.filterNot { isPreview(it.id) }
        return (stable.ifEmpty { candidates }).maxByOrNull { it.id }
    }

    /** Lower tier = cheaper. Unknown names land mid-tier. */
    internal fun costTier(kind: ProviderKind, id: String): Int {
        val lower = id.lowercase()
        return when (kind) {
            ProviderKind.ANTHROPIC, ProviderKind.CLAUDE_ACCOUNT -> when {
                "haiku" in lower -> 0
                "sonnet" in lower -> 1
                "opus" in lower -> 2
                "fable" in lower || "mythos" in lower -> 3
                else -> 2
            }
            ProviderKind.OPENAI, ProviderKind.CODEX -> when {
                "nano" in lower -> 0
                "mini" in lower -> 1
                else -> 2
            }
            ProviderKind.GEMINI -> when {
                "flash" in lower && "lite" in lower -> 0
                "flash" in lower -> 1
                "pro" in lower -> 2
                else -> 3
            }
            ProviderKind.MISTRAL -> when {
                "ministral" in lower || "tiny" in lower -> 0
                "nemo" in lower || "small" in lower -> 1
                "medium" in lower -> 2
                "large" in lower -> 3
                else -> 2
            }
            ProviderKind.XAI -> when {
                "mini" in lower -> 0
                "fast" in lower -> 1
                else -> 2
            }
            ProviderKind.DEEPSEEK -> if ("chat" in lower) 0 else 1
            ProviderKind.LOCAL, ProviderKind.OLLAMA -> 0
        }
    }

    internal fun isPreview(id: String): Boolean {
        val lower = id.lowercase()
        return "preview" in lower || "-exp" in lower || "latest" in lower
    }
}
